lib/python/qmk/cli/new/keyboard.py

```diff
@@ -0,0 +1,182 @@
+"""This script automates creating the files necessary for a new QMK keyboard.
+"""
+import shutil
+import hashlib
+import re
+
+from milc import cli
+from pathlib import Path
+from datetime import datetime
+
+
+@cli.argument('-kb', '--keyboard', help='Specify keyboard name. Example: clueboard/66')
+@cli.argument('-pr', '--project', help='The human-friendly name of the keyboard. Example: Clueboard 66%')
+@cli.argument('-mcu', '--microcontroller', help='Specify the microcontroller used for the keyboard.')
+@cli.argument('-u', '--username', help='Your GitHub username. Will be pasted into generated files.')
+@cli.subcommand('Creates a new keyboard project.')
+
+def new_keyboard(cli):
+    """Creates a new keyboard project.
+    """
+
+    # Root path for template files
+    template_root_path = Path("quantum/template")
+    template_base_path = Path(template_root_path) / "base"
+    year = datetime.now().year
+
+    # valid microcontroller choices
+    mcus = (
+        ( 'at90usb1286', 'avr' ),
+        ( 'at90usb646' , 'avr' ),
+        ( 'atmega16u2' , 'avr' ),
+        ( 'atmega16u4' , 'avr' ),
+        ( 'atmega328p' , 'avr' ),
+        ( 'atmega32a'  , 'ps2avrgb' ),
+        ( 'atmega32u2' , 'avr' ),
+        ( 'atmega32u4' , 'avr' ),
+        ( 'STM32F042'  , 'stm32' ),
+        ( 'STM32F072'  , 'stm32' ),
+        ( 'STM32F103'  , 'stm32' ),
+        ( 'STM32F303'  , 'stm32' ),
+    )
+
+    # Generate the USB Product ID
+    def generate_pid(str):
+        str = str.encode('utf-8')
+        str = hashlib.sha1(str).hexdigest()[0:4].upper()
+        return str
+
+    # Rewrites the %YEAR%, %YOUR_NAME%, %KEYBOARD% and %PID% placeholders in the
+    #   new files.
+    def rewrite_source(file):
+        rw_file = Path(keyboard_path) / file
+        file_contents = rw_file.read_text() \
+            .replace("%YEAR%", str(year)) \
+            .replace("%YOUR_NAME%", user_name) \
+            .replace("%KEYBOARD%", keyboard) \
+            .replace("%FINAL_DIR%", final_directory) \
+            .replace("%KEYBOARD_NAME%", keyboard_name) \
+            .replace("%PID%", pid) \
+            .replace("%MCU%", mcu)
+        rw_file.write_text(file_contents)
+
+    def print_microcontrollers():
```

Comment:
There isn't a lot of value here. I'd probably get rid of this function and put the code inline.

Reply:
It was way longer when I wrote it the first time...